File: ag-ui/integrations/langgraph/typescript/src/middlewares/state-streaming.ts
/**
 * Custom middleware helpers for ag-ui LangGraph agents.
 */

import { createMiddleware } from "langchain";
import { BaseMessage, ToolMessage } from "@langchain/core/messages";

export interface StateItem {
  stateKey: string;
  tool: string;
  toolArgument: string;
}

/** Identity helper — exists purely for IDE type inference on the object literal. */
export const stateItem = (item: StateItem): StateItem => item;

/**
 * Middleware that injects `predict_state` metadata into model invocations so
 * that every `on_chat_model_stream` event carries it.
 *
 * Approach: wrap `request.model` with `model.withConfig({ metadata: {
 * predict_state } })` before passing it to the base handler. When the base
 * handler subsequently calls `bindTools()` on this RunnableBinding,
 * `_simpleBindTools` detects the RunnableBinding wrapper and creates a new
 * RunnableBinding that **preserves our config**. `RunnableBinding.invoke()`
 * then uses `mergeConfigs()` (which deep-merges metadata) to combine our
 * bound config with the LangGraph execution config, so `predict_state`
 * survives into every streaming event.
 */
export const stateStreamingMiddleware = (...items: StateItem[]) => {
  const predictState = items.map((i) => ({
    state_key: i.stateKey,
    tool: i.tool,
    tool_argument: i.toolArgument,
  }));

  const trackedTools = new Set(items.map((i) => i.tool));

  /**
   * Return true if intermediate state should be injected for this model call.
   *
   * Suppress only when the last tool that ran is one we track. If we injected
   * again after a tracked tool, predict_state would stream a second time for
   * that same tool on its next invocation — a true duplicate stream. Untracked
   * tools (e.g. open_canvas) are safe to inject after because the next call
   * may need to stream a tracked tool.
   */
  const isPreToolCall = (request: { messages?: BaseMessage[] }): boolean => {
    const msgs = request?.messages ?? [];
    if (msgs.length === 0) return true;
    const last = msgs[msgs.length - 1];
    if (!(last instanceof ToolMessage)) return true;
    return !trackedTools.has(last.name ?? "");
  };

  return createMiddleware({
    name: "StateStreamingMiddleware",
    wrapModelCall: async (request, handler) => {
      if (!isPreToolCall(request)) {
        return handler(request);
      }
      const modelWithState = (request.model as any).withConfig({
        metadata: { predict_state: predictState },
      });

      const m = request.model as any;
      const proto = Object.getPrototypeOf(m);
      const origCombine = proto._combineCallOptions;
      if (origCombine) {
        proto._combineCallOptions = function (options: any) {
          const combined = origCombine.call(this, options);
          combined.metadata = {
            ...(this.defaultOptions?.metadata ?? {}),
            ...(options?.metadata ?? {}),
            predict_state: predictState,
          };
          return combined;
        };
      }
      try {
        return await handler({ ...request, model: modelWithState });
      } finally {
        if (origCombine) proto._combineCallOptions = origCombine;
      }
    },
  });
};

File: ag-ui/integrations/langgraph/python/ag_ui_langgraph/middlewares/state_streaming.py
"""
Custom middleware helpers for ag-ui LangGraph agents.
"""
from dataclasses import dataclass
from typing import Any, Awaitable, Callable

from langchain.agents.middleware import AgentMiddleware, ModelRequest
from langchain_core.messages import ToolMessage
from langchain_core.runnables.config import ensure_config, var_child_runnable_config


def _with_intermediate_state(config: dict, emit_intermediate_state: list) -> dict:
    metadata = {**config.get("metadata", {}), "predict_state": emit_intermediate_state}
    return {**config, "metadata": metadata}


@dataclass(frozen=True)
class StateItem:
    state_key: str
    tool: str
    tool_argument: str

class StateStreamingMiddleware(AgentMiddleware):
    def __init__(self, *items: StateItem) -> None:
        self._emit_intermediate_state = [
            {"state_key": i.state_key, "tool": i.tool, "tool_argument": i.tool_argument}
            for i in items
        ]

    def _is_pre_tool_call(self, request: ModelRequest) -> bool:
        """Return True if this model call precedes a tool call for the current turn.

        When the last message is a ToolMessage the tool has already run and the
        model is being called for a follow-up response.  Injecting
        emit_intermediate_state in that case causes predict_state streaming to
        fire again if the model decides to call the same tool a second time,
        producing a duplicate stream.
        """
        msgs = request.messages
        if not (msgs and isinstance(msgs[-1], ToolMessage)):
            return True
        # Only suppress if the last tool is one we're actually tracking
        # (prevents duplicate stream if the same tool is called again)
        last_tool_name = getattr(msgs[-1], 'name', None)
        tracked_tools = {item["tool"] for item in self._emit_intermediate_state}
        return last_tool_name not in tracked_tools

    def wrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], Any],
    ) -> Any:
        if not self._is_pre_tool_call(request):
            return handler(request)
        config = _with_intermediate_state(ensure_config(), self._emit_intermediate_state)
        token = var_child_runnable_config.set(config)
        try:
            return handler(request)
        finally:
            var_child_runnable_config.reset(token)

    async def awrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], Awaitable[Any]],
    ) -> Any:
        if not self._is_pre_tool_call(request):
            return await handler(request)
        config = _with_intermediate_state(ensure_config(), self._emit_intermediate_state)
        token = var_child_runnable_config.set(config)
        try:
            return await handler(request)
        finally:
            var_child_runnable_config.reset(token)
